ranndy/models/vae.py

```diff
@@ -0,0 +1,76 @@
+# Based on: https://arxiv.org/pdf/1511.06349.pdf
+# TODO: Word dropout and historyless decoding
+
+from models.auto_encoder import SentenceAutoEncoder
+import tensorflow as tf
+
+
+class SentenceVAE(SentenceAutoEncoder):
+
+    def __init__(self, data_iterator, mode=tf.estimator.ModeKeys.TRAIN):
+
+        # Instance variables associated with VAE
+        self.encoding_mean = None
+        self.encoding_log_var = None # Note: Log(variance) is learned to ensure positivity
+        self.sampled_state = None
+        self.reconstruction_loss = None
+        self.kl_loss = None
+        self.kl_weight_scaling = 10 # TODO: Tune these values
+        self.kl_weight_offset = 5000
+
+        super().__init__(data_iterator, mode)
+
+    def _build_encoder(self):
+        self.encoder_cell = tf.nn.rnn_cell.LSTMCell(self.lstm_size, name="vae_encoder_cell")
+        output, encoder_state_tuple = tf.nn.dynamic_rnn(self.encoder_cell, self.enc_embedding_output,
+                                             sequence_length=self.iterator.source_length, dtype=tf.float32)
+
+        # Since encoder state is an LSTMStateTuple, we extract only the state
+        encoder_state = encoder_state_tuple.c
+
+        # Layers to convert output to mean and variance
+        #   Note: Paper uses linear layer (i.e. no activation)
+        self.encoding_mean = tf.layers.dense(encoder_state, self.lstm_size, activation=None, name="linear_mean_layer")
+        self.encoding_log_var = tf.layers.dense(encoder_state, self.lstm_size, activation=None, name="linear_var_layer")
+
+        # Set decoder state input as random sample from generated decoder distribution
+        self.sampled_state = self.gaussian_sample(self.encoding_mean, tf.exp(self.encoding_log_var))
+        self.decoder_state_input = tf.contrib.rnn.LSTMStateTuple(c=self.sampled_state, h=encoder_state_tuple.h)
+
+    def _build_trainer(self):
+
+        # Build reconstruction loss
+        target_weights = tf.sequence_mask(self.iterator.target_length, dtype=self.logits.dtype)
+        crossent = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.iterator.target_output,
+                                                                  logits=self.logits)
+        self.reconstruction_loss = tf.reduce_sum(crossent * target_weights)/self.batch_size
+
```

Comment:
If this code is identical to base AE code, use a super() call.

Reply:
The code is not the same, calling super() would build an incorrect trainer which doesn't account for the KL and reconstruction loss. I will pull out the shared part into a utility function to reduce repeated code.